src/library/dao/trackdao.cpp

```diff
@@ -1622,14 +1622,16 @@ void TrackDAO::markTrackLocationsAsDeleted(const QString& directory) {
     }
 }
 
-// Look for moved files. Look for files that have been marked as "deleted on disk"
-// and see if another "file" with the same name and filesize exists in the track_locations
-// table. That means the file has moved instead of being deleted outright, and so
-// we can salvage your existing metadata that you have in your DB (like cue points, etc.).
-void TrackDAO::detectMovedFiles(QSet<int>* pTracksMovedSetOld, QSet<int>* pTracksMovedSetNew) {
-    //This function should not start a transaction on it's own!
-    //When it's called from libraryscanner.cpp, there already is a transaction
-    //started!
+// Look for moved files. Look for files that have been marked as
+// "deleted on disk" and see if another "file" with the same name and
+// files size exists in the track_locations table. That means the file has
+// moved instead of being deleted outright, and so we can salvage your
+// existing metadata that you have in your DB (like cue points, etc.).
+bool TrackDAO::detectMovedFiles(QSet<int>* pTracksMovedSetOld,
+        QSet<int>* pTracksMovedSetNew, volatile const bool* pCancel) {
```

Comment:
But then we should rather improve this function. Currently canceling means that only minimal progress is lost.
If this is canceled after 50 seconds run time in your scenario all of the progress we made so far will be lost. 
Coming to think of it. I also would expect that a canceled scan does a proper clean up. 

Your solution might fix the long shutdown time of mixxx in canceled scans but I don't think some one would rescan his entire library during a performance. The better solution in the long run is to improve the performance of this.

It should also be possible to write a unit-test for the clean-up functions. They only depend on internal library state.